### PR TITLE
Use `ExprRewriter` in `pre_cast_lit_in_comparison`

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3259

 # Rationale for this change

This change makes the optimizer support nested expressions, including aliased expressions.

# What changes are included in this PR?

Use `ExprRewriter`.

# Are there any user-facing changes?

More optimizations!

### CODE REVIEW

**Comment by liukun4515** [comment]
nice pr.
I am ready to do this refactor using the trait `ExprRewriter`, you got ahead.

**Review by liukun4515 — APPROVED** [review]
LGTM thanks @andygrove

File: datafusion/optimizer/src/pre_cast_lit_in_comparison.rs [hunk 1]
```diff
@@ -318,6 +317,20 @@ mod tests {
         assert_eq!(optimize_test(expr_lt, &schema), expected);
     }
 
+    #[test]
+    fn nested() {
```
Comment (liukun4515) [inline datafusion/optimizer/src/pre_cast_lit_in_comparison.rs:321]: Does this test require this pr to pass?
I think the original logic can handle this case

Comment (andygrove) [reply datafusion/optimizer/src/pre_cast_lit_in_comparison.rs:321]: No, this is just an additional test to make sure that this PR did not cause any regressions.

**Comment by andygrove** [comment]
Thanks for the review @liukun4515

**Pull request merged**